Fix #16015 regression in V12: if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing correctly in Sales Order

zuiko opened the pull request:
#Fix #16015 regression in V12
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order 
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed  as VAT codes in dictionnary.

zuiko committed:
Update functions.lib.php

Fix #16015 regression in V12
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order 
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed  as VAT codes in dictionnary.

zuiko committed:
Update functions.lib.php

#Fix #16015 regression in V12 development branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed as VAT codes in dictionnary.

zuiko committed:
Update functions.lib.php

#Fix #16015 regression in V12 development branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed as VAT codes in dictionnary.

zuiko committed:
Update functions.lib.php

#Fix #16015 regression in V12 development branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed as VAT codes in dictionnary.

eldy commented:
Allowing special characters into a "code" (that is a key to identify uniquely a vat rate) does not seems to be a good idea.
Why not just sanitizing the code with value "dol_string_nospecial()"

zuiko committed:
Update functions.lib.php

#Fix #16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
code enhancement if $amount is not numeric (about remaining + and -)
The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
Test " (" existence before elimination of the taxe code string

zuiko commented:
Hi @eldy 
the problem is that's a regression in V12.0.4 for many users using taxes codes with some characters allowed without problems before and rejected now with not obvious errors or strange behaviour in updates around product quantities or prices.
Sometimes it's even the case for natives TVA presets as 85, 85NPR, 85NPROM, 85NPROMOMR which begin by numerical characters.
So, even if we don't clean everything as I tried to do if is not numeric, we have at least to cut the string from the taxe code as " (taxe code)" for which it could be legitimate for preceding users to use their proper ancient codes even in macedonian, bulgarian or greek etc... without sql errors.

Putting at least this code:
```
$codepos = strpos($amount, ' (' );//if the string contains " (" it contains a code name, so eliminate the end of string including parenthesis, so 85NPR VAT code will work
if ($codepos > 0) {
	$amount = substr($amount, 0, $codepos);
}
```
after:
`if (!is_numeric($amount)) {`
and before:
`$amount = preg_replace('/[a-zA-Z\/\\\*\(\)\<\>]/', '', $amount);`
(without even adding "_" in the replace)

seems mandatory to me.

May be there is also a lack of control for taxes codes in the dictionnary...

What do you think about this issue?

eldy commented:
If problem occurs due to a tax code, I don't think the trouble is inside the method price2num. This method is designed to be able to convert a number entered by a suer into a php number. We should not have special char and if we have, it means user try to hack the system.
I think the real trouble is the code that call price2num of a string that is not coming from a field that should be an amount, quantity or numeric value.
If a price2num is called on a var that contains "10 (VATCODE123)", it means there is a small code that miss. 
If we look at different places of code, we see that we have on a VAT code this king of cleaning:

```
$tmpvat = price2num(preg_replace('/\s*\(.*\)/', '', $tva_tx));
or
$tva_tx = preg_replace('/\s*\(.*\)/', '', $tva_tx); // Remove code into vatrate.
```

So if we have still trouble due to VATCODE with special char, it means we are missing this replacement before calling the price2num (The goal of price2num is not to keep safe special characters, it is to convert a number with a decimal value that depends on locale of user into a number that use . as decimal value).
So for me changing the price2num is not a good solution. We must fix the page that call price2num of a Vat code. Method is designed to be called on the rate inside the Vat code string.

zuiko commented:
I undestand and agree with what you write, the issue is upstream of price2num but I fixed the issue here because of the pre-existing code in it that I saw:
`if (!is_numeric($amount)) { `
This code has not been created by me and it has been put here for some reason, I don't know.
So, I believed that receiving here some no numeric value (as TVA codes) could be considered as normal for some reason...

In other hand, I confirm by my logs implemented for debugging that price2num receives strings like "10 (VATCODE123)"...

It could be useful to trap the caller when the condition `(!is_numeric($amount))` is true.

In complement I confirm that the code 
`$tva_tx = preg_replace('/\s*\(.*\)/', '', $tva_tx);`
seems able to remove `" (VATCODE123)"` even with special characters in it, as my code do, so there is a lack of preg_replace for VAT rate somewhere.
And I know that finding a lack of doing something in the code is the worst thing to find ;-)

In /htdocs/commande/card.php there are 26 calls to price2num and I didn't find nothing strange...

zuiko commented:
@eldy 
for customer order line editing the guilty caller is TADA...:
```
			$txtva = price2num($txtva);
			$txlocaltax1 = price2num($txlocaltax1);
			$txlocaltax2 = price2num($txlocaltax2);
```
in /htdocs/commande/class/commande.class.php lines 3123-3125 (I put a trap on the code),
and there  is also a potential problem for local taxes (I red something on forum about it).
But if you grep with "price2num($txtva" you will see many files to verify:
propal.class.php
commande.class.php
facture-rec.class.php
facture.class.php
contrat.class.php
fichinterrec.class.php
fournisseur.commande.class.php
fournisseur.facture.class.php
supplier_proposal.class.php
some have a preceding cleaning, for VAT, some not, and it's never the case for local taxes, I fear.
So I wonder if it would not be better to centralise the cleaning instead of hunting the forgotten cleaning somewhere. 
Even we are able to find all today, it is not guaranted that some others will appear in the future.

And here I was speaking only about VAT and local taxes but if you analyse this line:
`$tmpamounttoshow = price2num($fromcompany->capital); // This field is a free string`

For me it's another dubious call in /htdocs/core/lib/pdf.lib.php line 992
Because if you set $tmpamounttoshow to  "100 000 €"  price2num will currently not eliminate €

So verifying the 2093 calls to price2num in 250 concerned files could be time consuming...
For me, a so important function has to defend itself from bad calls and, if possible, silently.
But my solution could not be perfect and has to be challenged, re-read and tested deeply for a so much called function.
Notice: for is_numeric PHP function a number with a comma as decimal separator is never considered as numeric, so in many cases (in France etc...) (!is_numeric($amount)) is near allways true.

zuiko committed:
Update functions.lib.php

Improve defence in price2num in order to at least
#Fix #16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
- The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
- Test "(taxrate)" existence before elimination of the taxe code string
- It works also for local taxes
- try to clean all others not numeric strings passed to price2num
- propose a log for analysis if it does not succeed to found a numerical value

zuiko committed:
Update functions.lib.php

Improve defence in price2num in order to at least
#Fix #16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
- The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
- Test "(taxrate)" existence before elimination of the taxe code string
- It works also for local taxes
- try to clean all others not numeric strings passed to price2num
- propose a log for analysis if it does not succeed to found a numerical value

zuiko committed:
Update functions.lib.php

Improve defence in price2num in order to at least
#Fix #16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
- The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
- Test "(taxrate)" existence before elimination of the taxe code string
- It works also for local taxes
- try to clean all others not numeric strings passed to price2num
- propose a log for analysis if it does not succeed to found a numerical value

zuiko commented:
@eldy
On my last commit I proposed a better "central" defence for price2num.
Before testing if string is numeric, I begin to execute the most common code which could have been forgotten:
```
$amount = preg_replace('/\s*\(.*\)/', '', $amount); // Remove forgotten code into vatrate or local taxe.
$amount = preg_replace('/\s*/', '', $amount);// Remove remaining spaces if any
```
If after this, the string is allways not numeric, I procede to the cleaning as a complete defence of upstream lack of string treatment or typing errors.

The only characters present at the end of my code if at first (!is_numeric($amount)) for $amount  can be:

- some digits 0 to 9 
- two minus (-) maximum (one for principal sign, one for exposure sign)
- one dot (.) maximum
- one E maximum (for exposure)

that's all

So this avoids all sql errors coming from upstream lack of string treatment or typing errors.
At the end of my treatment, if $amount is definively not numeric (which is possible because I don't order the remaining characters), I propose a log with level LOG_ERR with some infos to debug and finally I execute $amount = ''; so 0 will be passed as numrical value to the database.

**NOTICE:**
I would have preferred to show a soft error message to the user such "Typing error or wrong argument passing by an internal function" and leave the original value unchanged but I didn't succeed to do that.
For example `dol_print_error($db,texte);` seems to "heavy" for such a simple message and I did not find anything else.
I didn't find how to let the original value inchanged in price2num.
**NOTICE END**

Typing values is not restricted, for example -12,5e2 is supported as before and will be shown as -1 250
and also -12,5e-2 as -0,125

This could offer also, now, an opportunity to delete repetitives string treatments about VAT codes which are not completly homogeneous...

I did not change nothing at the end of the function written to adapt number presentation to local settings.

zuiko committed:
Improve defence on price2num

Improve defence in price2num in order to at least
#Fix Dolibarr#16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
- The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
- Test "(taxrate)" existence before elimination of the taxe code string
- It works also for local taxes
- try to clean all others not numeric strings passed to price2num
- propose a log for analysis if it does not succeed to found a numerical value

eldy commented:
You should have no problem with localtax because locatax rate are never represented with "rate (code)". This occurs only on main vat when data is coming from a user input from a combo box to select vat.
So localtax are always get with a value "rate"

As i said, we should not have to change something into internal function price2num. If a bad data is send to price2num by buggy code, we must fix the root cause (so the caller) and not change the goal of the function price2num as this as bad side effect.

zuiko committed:
Ultimate proposal to improve defence in price2num

Improve defence in price2num in order to at least
#Fix #16015 regression in V12 branch
if VAT code is written in non latin characters or begins by numerical characters, unable to edit the pricing in Sales Order
- The consequence of this code is that all characters are allowed as VAT codes in dictionnary.
- Test "(taxrate)" existence before elimination of the taxe code string
- Avoid to enter the first (!is_numeric($amount) execution code if $amount=='' which is normal and often and with nothing to do
- It works also for any future cases
- try to clean all others not numeric strings passed to price2num
- propose a log for analysis if it does not succeed to found a numerical value
- propose a javascript alert for typing errors (I don't know if it's an approved behaviour in Dolibarr in such a case).

zuiko commented:
OK @eldy
I understand what you say, 
in fact the regression for users appeared when price2num evolved between V12.0.3 and V12.0.4 where  the 1st test  on (!is_numeric($amount) as been introduced, I don't know why., there are some commits touching it.
So, you are right, modifying price2num could have bad side effects. We saw it at first for #15820 and here for #16015...

On the occasion of this regression, in fact, my concern was just to improve this code recently introduced, 
If you compare the 2 versions, the main difference is this code:
```
if (!is_numeric($amount)) {
			$amount = preg_replace('/[a-zA-Z\/\\\*\(\)\<\>]/', '', $amount);
		}

		if ($option == 2 && $thousand == '.' && preg_match('/\.(\d\d\d)$/', (string) $amount)) {	// It means the . is used as a thousand separator and string come frominput data, so 1.123 is 1123
			$amount = str_replace($thousand, '', $amount);
		}
```
And I continue to think that this code is not complete.
For example, it is executed it if  `$amount == ''` which is often but useless...

An other example, for the official V12.0.4 version of price2num :
If in a field quantity in customer order you type:
**10,5e-2**  the field will show currently **8,5** (it's a little bit difficult to understand this result)
**10.5e-2** the field will show currently  **0,105** (which is correct)

For my last version, the result is the same in the field: **0,105**

So I will propose in an other personal branch upstream corrections to fix this precise regression about VAT cyrillic taxe but I fear we  will have to improve price2num code a day, because it seems not perfect currently.

For your information, my patch is currently in production on my installation for my users and I'm waiting for their alerts...

EDIT:
I didn't see that you covered the upstream fix by:
[FIX Bad cleaning of VAT rate when numbers are into code](https://github.com/Dolibarr/dolibarr/commit/fd4dab495f010fc170c94ab3e6d5a47be7e7f605)

eldy commented:
Seems already fixed.
